extensions/librdkafka/ConsumeKafka.cpp

```diff
@@ -0,0 +1,553 @@
+  }
+  std::vector<std::string> matching_headers;
+  for (std::size_t header_idx = 0;; ++header_idx) {
+    const char* value;  // Not to be freed
+    std::size_t size;
+    if (RD_KAFKA_RESP_ERR_NO_ERROR != rd_kafka_header_get(headers_raw, header_idx, header_name.c_str(), (const void**)(&value), &size)) {
+      break;
+    }
+    if (size < std::numeric_limits<int>::max()) {
+      logger_->log_debug("%.*s", static_cast<int>(size), value);
+    }
+    matching_headers.emplace_back(value, size);
+  }
+  return matching_headers;
+}
+
+std::vector<std::pair<std::string, std::string>> ConsumeKafka::get_flowfile_attributes_from_message_header(const rd_kafka_message_t* message) const {
+  std::vector<std::pair<std::string, std::string>> attributes_from_headers;
+  for (const std::string& header_name : headers_to_add_as_attributes_) {
+    const std::vector<std::string> matching_headers = get_matching_headers(message, header_name);
+    if (matching_headers.size()) {
+      attributes_from_headers.emplace_back(header_name, utils::get_encoded_string(resolve_duplicate_headers(matching_headers), message_header_encoding_attr_to_enum()));
+    }
+  }
+  return attributes_from_headers;
+}
+
+class WriteCallback : public OutputStreamCallback {
+ public:
+  WriteCallback(char *data, uint64_t size) :
+      data_(reinterpret_cast<uint8_t*>(data)),
+      dataSize_(size) {}
+  uint8_t* data_;
+  uint64_t dataSize_;
+  int64_t process(const std::shared_ptr<io::BaseStream>& stream) {
+    int64_t ret = 0;
+    if (data_ && dataSize_ > 0)
+      ret = stream->write(data_, dataSize_);
+    return ret;
+  }
+};
+
+void ConsumeKafka::add_kafka_attributes_to_flowfile(std::shared_ptr<FlowFileRecord>& flow_file, const rd_kafka_message_t* message) const {
+  // We do not currently support batching messages into a single flowfile
+  flow_file->setAttribute(KAFKA_COUNT_ATTR, "1");
+  const utils::optional<std::string> message_key = utils::get_encoded_message_key(message, key_attr_encoding_attr_to_enum());
+  if (message_key) {
+    flow_file->setAttribute(KAFKA_MESSAGE_KEY_ATTR, message_key.value());
+  }
+  flow_file->setAttribute(KAFKA_OFFSET_ATTR, std::to_string(message->offset));
+  flow_file->setAttribute(KAFKA_PARTITION_ATTR, std::to_string(message->partition));
+  flow_file->setAttribute(KAFKA_TOPIC_ATTR, rd_kafka_topic_name(message->rkt));
+}
+
+std::vector<std::shared_ptr<FlowFileRecord>> ConsumeKafka::transform_messages_into_flowfiles(
+    const std::vector<std::unique_ptr<rd_kafka_message_t, utils::rd_kafka_message_deleter>>& messages, core::ProcessSession* session) const {
+  std::vector<std::shared_ptr<FlowFileRecord>> flow_files_created;
+  for (const auto& message : messages) {
+    std::string message_content = extract_message(message.get());
+    if (message_content.empty()) {
+      logger_->log_debug("Error: message received contains no data.");
+      return {};
```

Comment:
Should this be a `continue` instead of a `return {}`?  If we find an empty message, I'm not sure we want to discard all other messages before and after it.  (Probably at line 515, too.)

Reply:
Updated this line to `continue`. Doing the same thing with line 515 is unsafe, because due to message demarcation, we can be halfway through a message when the failure happens.

Reply:
Update:
Reworked this part. Now I/O failures are handled by storing non-written messages and attempting to write them on the start of the next onTrigger.